File: parallel-algorithms/sorting/parallel_merge_sort.cpp
/*
 * Parallel Merge Sort Implementation in C++
 *
 * This implementation uses OpenMP for parallel execution, providing:
 * - Low-level control over parallelism
 * - Compiler directives for parallelization
 * - Portable across platforms
 * - Excellent performance
 *
 * Features:
 * - OpenMP parallel sections and tasks
 * - Thread-safe operations
 * - Atomic operations for counters
 * - RAII for resource management
 * - Modern C++17/20 features
 *
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 *
 * To compile and run:
 *   g++ -std=c++17 -fopenmp -O3 parallel_merge_sort.cpp -o parallel_merge_sort
 *   ./parallel_merge_sort
 *
 * Or with clang:
 *   clang++ -std=c++17 -fopenmp -O3 parallel_merge_sort.cpp -o parallel_merge_sort
 *
 * Author: Algorithms Multiverse
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>
#include <iomanip>
#include <atomic>
#include <string>
#include <map>
#include <cmath>

#ifdef _OPENMP
#include <omp.h>
#endif

// Result structure for sorting operations
struct SortResult {
    std::vector<int> sortedArray;
    double timeTaken;  // seconds
    long long comparisons;
    std::string method;
    int numThreads;
};

class ParallelMergeSort {
private:
    static constexpr int SEQUENTIAL_THRESHOLD = 1000;
    static constexpr int PROCESS_THRESHOLD = 100000;

    int numThreads;
    std::atomic<long long> comparisons;

    // Merge two sorted vectors
    std::vector<int> merge(const std::vector<int>& left,
                          const std::vector<int>& right,
                          bool countComparisons = true) {
        std::vector<int> result;
        result.reserve(left.size() + right.size());

        size_t i = 0, j = 0;

        while (i < left.size() && j < right.size()) {
            if (countComparisons) {
                comparisons.fetch_add(1, std::memory_order_relaxed);
            }

            if (left[i] <= right[j]) {
                result.push_back(left[i++]);
            } else {
                result.push_back(right[j++]);
            }
        }

        result.insert(result.end(), left.begin() + i, left.end());
        result.insert(result.end(), right.begin() + j, right.end());

        return result;
    }

public:
    ParallelMergeSort(int threads = 0)
        : numThreads(threads > 0 ? threads : omp_get_max_threads()),
          comparisons(0) {
#ifdef _OPENMP
        omp_set_num_threads(numThreads);
#endif
    }

    // Sequential merge sort for base cases
    std::vector<int> sequentialMergeSort(const std::vector<int>& arr) {
        if (arr.size() <= 1) {
            return arr;
        }

        size_t mid = arr.size() / 2;
        std::vector<int> left(arr.begin(), arr.begin() + mid);
        std::vector<int> right(arr.begin() + mid, arr.end());

        left = sequentialMergeSort(left);
        right = sequentialMergeSort(right);

        return merge(left, right);
    }

    // Parallel merge sort using OpenMP tasks
    std::vector<int> parallelMergeSortTasks(const std::vector<int>& arr, int depth = 0) {
        if (arr.size() <= SEQUENTIAL_THRESHOLD) {
            return sequentialMergeSort(arr);
        }

        // Limit parallel depth to prevent task explosion
        int maxDepth = static_cast<int>(std::log2(numThreads));
        if (depth >= maxDepth) {
            return sequentialMergeSort(arr);
        }

        size_t mid = arr.size() / 2;
        std::vector<int> left(arr.begin(), arr.begin() + mid);
        std::vector<int> right(arr.begin() + mid, arr.end());

        std::vector<int> leftResult, rightResult;

#ifdef _OPENMP
        #pragma omp parallel sections
        {
            #pragma omp section
            {
                leftResult = parallelMergeSortTasks(left, depth + 1);
            }
            #pragma omp section
            {
                rightResult = parallelMergeSortTasks(right, depth + 1);
            }
        }
#else
        leftResult = parallelMergeSortTasks(left, depth + 1);
        rightResult = parallelMergeSortTasks(right, depth + 1);
#endif

        return merge(leftResult, rightResult);
    }

    // Parallel merge sort using OpenMP task directive (more efficient)
    std::vector<int> parallelMergeSortTaskDirective(const std::vector<int>& arr, int depth = 0) {
        if (arr.size() <= SEQUENTIAL_THRESHOLD) {
            return sequentialMergeSort(arr);
        }

        int maxDepth = static_cast<int>(std::log2(numThreads));
        if (depth >= maxDepth) {
            return sequentialMergeSort(arr);
        }

        size_t mid = arr.size() / 2;
        std::vector<int> left(arr.begin(), arr.begin() + mid);
        std::vector<int> right(arr.begin() + mid, arr.end());

        std::vector<int> leftResult, rightResult;

#ifdef _OPENMP
        #pragma omp task shared(leftResult) if(depth < maxDepth)
        leftResult = parallelMergeSortTaskDirective(left, depth + 1);

        #pragma omp task shared(rightResult) if(depth < maxDepth)
        rightResult = parallelMergeSortTaskDirective(right, depth + 1);

        #pragma omp taskwait
#else
        leftResult = parallelMergeSortTaskDirective(left, depth + 1);
        rightResult = parallelMergeSortTaskDirective(right, depth + 1);
#endif

        return merge(leftResult, rightResult);
    }

    // Wrapper for task directive version
    std::vector<int> parallelMergeSortTaskWrapper(const std::vector<int>& arr) {
        std::vector<int> result;
#ifdef _OPENMP
        #pragma omp parallel
        {
            #pragma omp single
            result = parallelMergeSortTaskDirective(arr, 0);
        }
#else
        result = parallelMergeSortTaskDirective(arr, 0);
#endif
        return result;
    }

    // Parallel merge sort using chunking strategy
    std::vector<int> parallelMergeSortChunked(const std::vector<int>& arr) {
        if (arr.size() <= SEQUENTIAL_THRESHOLD) {
            return sequentialMergeSort(arr);
        }

        size_t chunkSize = std::max(arr.size() / numThreads,
                                    static_cast<size_t>(SEQUENTIAL_THRESHOLD));

        std::vector<std::vector<int>> chunks;
        for (size_t i = 0; i < arr.size(); i += chunkSize) {
            size_t end = std::min(i + chunkSize, arr.size());
            chunks.emplace_back(arr.begin() + i, arr.begin() + end);
        }

        // Sort chunks in parallel
        std::vector<std::vector<int>> sortedChunks(chunks.size());

#ifdef _OPENMP
        #pragma omp parallel for
#endif
        for (size_t i = 0; i < chunks.size(); ++i) {
            sortedChunks[i] = sequentialMergeSort(chunks[i]);
        }

        // Merge sorted chunks
        while (sortedChunks.size() > 1) {
            std::vector<std::vector<int>> merged;

            for (size_t i = 0; i < sortedChunks.size(); i += 2) {
                if (i + 1 < sortedChunks.size()) {
                    merged.push_back(merge(sortedChunks[i], sortedChunks[i + 1], false));
                } else {
                    merged.push_back(sortedChunks[i]);
                }
            }

            sortedChunks = std::move(merged);
        }

        return sortedChunks.empty() ? std::vector<int>() : sortedChunks[0];
    }

    // Adaptive parallel merge sort
    std::vector<int> adaptiveParallelMergeSort(const std::vector<int>& arr) {
        size_t n = arr.size();

        if (n <= SEQUENTIAL_THRESHOLD) {
            return sequentialMergeSort(arr);
        } else if (n <= PROCESS_THRESHOLD) {
            return parallelMergeSortTaskWrapper(arr);
        } else {
            return parallelMergeSortChunked(arr);
        }
    }

    // Sort with metrics
    SortResult sortWithMetrics(const std::vector<int>& arr, const std::string& method) {
        comparisons.store(0, std::memory_order_relaxed);

        auto start = std::chrono::high_resolution_clock::now();

        std::vector<int> sorted;
        if (method == "sequential") {
            sorted = sequentialMergeSort(arr);
        } else if (method == "tasks") {
            sorted = parallelMergeSortTasks(arr);
        } else if (method == "task_directive") {
            sorted = parallelMergeSortTaskWrapper(arr);
        } else if (method == "chunked") {
            sorted = parallelMergeSortChunked(arr);
        } else if (method == "adaptive") {
            sorted = adaptiveParallelMergeSort(arr);
        } else {
            sorted = adaptiveParallelMergeSort(arr);
        }

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> duration = end - start;

        return SortResult{
            sorted,
            duration.count(),
            comparisons.load(std::memory_order_relaxed),
            method,
            numThreads
        };
    }

    // Check if array is sorted
    static bool isSorted(const std::vector<int>& arr) {
        for (size_t i = 1; i < arr.size(); ++i) {
            if (arr[i] < arr[i - 1]) {
                return false;
            }
        }
        return true;
    }

    // Calculate speedup
    static double calculateSpeedup(double seqTime, double parallelTime) {
        return parallelTime > 0 ? seqTime / parallelTime : 0.0;
    }

    // Calculate efficiency
    static double calculateEfficiency(double speedup, int numThreads) {
        return numThreads > 0 ? speedup / numThreads : 0.0;
    }
};

// Benchmark all methods
std::map<std::string, SortResult> benchmarkAllMethods(const std::vector<int>& arr, int numThreads) {
    std::map<std::string, SortResult> results;
    std::vector<std::string> methods = {
        "sequential", "tasks", "task_directive", "chunked", "adaptive"
    };

    ParallelMergeSort sorter(numThreads);

    for (const auto& method : methods) {
        results[method] = sorter.sortWithMetrics(arr, method);
    }

    return results;
}

int main() {
    std::cout << std::string(80, '=') << "\n";
    std::cout << "PARALLEL MERGE SORT DEMONSTRATION (C++ with OpenMP)\n";
    std::cout << std::string(80, '=') << "\n";

#ifdef _OPENMP
    std::cout << "OpenMP is enabled\n";
    std::cout << "Number of available threads: " << omp_get_max_threads() << "\n";
#else
    std::cout << "OpenMP is NOT enabled (sequential execution only)\n";
#endif

    // Test with different sizes
    std::vector<int> sizes = {1000, 10000, 100000};

    for (int size : sizes) {
        std::cout << "\n" << std::string(80, '=') << "\n";
        std::cout << "Testing with " << size << " elements\n";
        std::cout << std::string(80, '=') << "\n";

        // Generate random array
        std::vector<int> arr(size);
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 100000);

        for (int& val : arr) {
            val = dis(gen);
        }

        // Benchmark all methods
        auto results = benchmarkAllMethods(arr, omp_get_max_threads());

        // Display results
        std::cout << "\n" << std::left << std::setw(18) << "Method"
                  << std::setw(15) << "Time (s)"
                  << std::setw(20) << "Comparisons"
                  << std::setw(10) << "Verified" << "\n";
        std::cout << std::string(80, '-') << "\n";

        double seqTime = results.count("sequential") ? results["sequential"].timeTaken : 0.0;

        for (const auto& [method, result] : results) {
            std::string verified = ParallelMergeSort::isSorted(result.sortedArray) ? "✓" : "✗";

            std::cout << std::left << std::setw(18) << method
                      << std::fixed << std::setprecision(6)
                      << std::setw(15) << result.timeTaken
                      << std::setw(20) << result.comparisons
                      << std::setw(10) << verified << "\n";

            if (method != "sequential" && seqTime > 0) {
                double speedup = ParallelMergeSort::calculateSpeedup(seqTime, result.timeTaken);
                double efficiency = ParallelMergeSort::calculateEfficiency(speedup, result.numThreads);
                std::cout << std::setw(18) << ""
                          << "Speedup: " << std::fixed << std::setprecision(2) << speedup << "x  "
                          << "Efficiency: " << std::fixed << std::setprecision(2)
                          << efficiency * 100 << "%\n";
            }
        }
    }

    std::cout << "\n" << std::string(80, '=') << "\n";
    std::cout << "OPENMP SCALABILITY ANALYSIS\n";
    std::cout << std::string(80, '=') << "\n";

    // Test with different thread counts
    std::cout << "\nScalability test (50,000 elements):\n";

    std::vector<int> testArr(50000);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 100000);
    for (int& val : testArr) {
        val = dis(gen);
    }

    std::cout << "\n" << std::left << std::setw(10) << "Threads"
              << std::setw(15) << "Time (s)"
              << std::setw(10) << "Speedup" << "\n";
    std::cout << std::string(40, '-') << "\n";

    double baselineTime = 0.0;
    int maxThreads = omp_get_max_threads();

    for (int threads = 1; threads <= maxThreads; ++threads) {
        ParallelMergeSort sorter(threads);
        auto result = sorter.sortWithMetrics(testArr, "task_directive");

        if (threads == 1) {
            baselineTime = result.timeTaken;
        }

        double speedup = ParallelMergeSort::calculateSpeedup(baselineTime, result.timeTaken);
        std::cout << std::left << std::setw(10) << threads
                  << std::fixed << std::setprecision(6) << std::setw(15) << result.timeTaken
                  << std::fixed << std::setprecision(2) << speedup << "x\n";
    }

    std::cout << "\n" << std::string(80, '=') << "\n";
    std::cout << "WHEN TO USE PARALLEL MERGE SORT IN C++ WITH OPENMP\n";
    std::cout << std::string(80, '=') << "\n";
    std::cout << R"(
Parallel merge sort with OpenMP is beneficial when:

✓ Dataset size > 10,000 elements
✓ Multiple CPU cores available
✓ CPU-bound comparison operations
✓ Low-level control needed

C++ + OpenMP advantages:
✓ Compiler directives (easy to parallelize)
✓ Low overhead (~100-200ns per task)
✓ Excellent performance (near-native)
✓ Portable across compilers (GCC, Clang, Intel, MSVC)
✓ Fine-grained control over threads
✓ No runtime overhead (compiled to native code)

Avoid parallel merge sort when:

✗ Small datasets (<1,000 elements)
✗ Single-core systems
✗ OpenMP not available
✗ Simple comparisons with negligible CPU cost

Strategy selection:
- Sequential: < 1,000 elements
- Tasks/Sections: 1,000 - 100,000 elements (recursive)
- Task Directive: Best performance (work stealing)
- Chunked: > 100,000 elements (better cache locality)
- Adaptive: Let the algorithm decide

Performance characteristics:
- Task creation: ~100-200ns overhead
- Work stealing: Built into OpenMP runtime
- Typical speedup: 3.5-3.8x on 4 cores
- Efficiency: 85-95% for optimal workload
- Cache effects: Significant for large arrays

Best practices:
1. Use #pragma omp task for recursive algorithms
2. Use #pragma omp parallel for for independent iterations
3. Limit task depth to prevent explosion
4. Use if clause to control parallelization
5. Compile with -O3 for optimization
6. Profile with perf or gprof

OpenMP directives used:
- #pragma omp parallel: Create parallel region
- #pragma omp sections: Divide work into sections
- #pragma omp task: Create asynchronous task
- #pragma omp taskwait: Wait for child tasks
- #pragma omp parallel for: Parallel loop

Memory considerations:
- Stack allocation preferred (faster)
- Avoid heap allocations in hot paths
- Consider cache line size (64 bytes)
- Use std::vector::reserve() to avoid reallocations

Compiler flags:
- GCC/Clang: -fopenmp -O3
- Intel: -qopenmp -O3
- MSVC: /openmp /O2

Environment variables:
- OMP_NUM_THREADS: Set number of threads
- OMP_SCHEDULE: Set scheduling policy
- OMP_PROC_BIND: Set thread affinity

Comparison with other approaches:
- OpenMP: Easiest to use, good performance
- TBB: Better abstractions, C++ friendly
- std::thread: Full control, more code
- std::async: High-level, limited control
)";

    std::cout << "\nDemonstration complete!\n";

    return 0;
}

File: parallel-algorithms/sorting/parallel_merge_sort.rs
/*
Parallel Merge Sort Implementation in Rust

This implementation uses Rayon for data parallelism, providing:
- Zero-cost abstractions
- Memory safety without garbage collection
- Work-stealing scheduler
- Fearless concurrency

Features:
- Rayon parallel iterators
- Thread pool configuration
- Lock-free operations
- Type safety and zero-cost abstractions

Time Complexity: O(n log n)
Space Complexity: O(n)

To compile and run:
1. Add to Cargo.toml:
   [dependencies]
   rayon = "1.7"

2. Run: cargo run --release

Author: Algorithms Multiverse
*/

use rayon::prelude::*;
use std::time::Instant;
use std::sync::atomic::{AtomicUsize, Ordering};

/// Result of a sorting operation with performance metrics
#[derive(Debug, Clone)]
pub struct SortResult {
    pub sorted_array: Vec<i32>,
    pub time_taken: std::time::Duration,
    pub comparisons: usize,
    pub method: String,
    pub num_threads: usize,
}

/// Parallel merge sort implementation
pub struct ParallelMergeSort {
    sequential_threshold: usize,
    comparisons: AtomicUsize,
}

impl ParallelMergeSort {
    /// Create a new ParallelMergeSort instance
    pub fn new(sequential_threshold: usize) -> Self {
        ParallelMergeSort {
            sequential_threshold,
            comparisons: AtomicUsize::new(0),
        }
    }

    /// Increment comparison counter (thread-safe)
    fn increment_comparisons(&self) {
        self.comparisons.fetch_add(1, Ordering::Relaxed);
    }

    /// Reset comparison counter
    fn reset_comparisons(&self) {
        self.comparisons.store(0, Ordering::Relaxed);
    }

    /// Get current comparison count
    fn get_comparisons(&self) -> usize {
        self.comparisons.load(Ordering::Relaxed)
    }

    /// Merge two sorted slices into one sorted vector
    fn merge(&self, left: &[i32], right: &[i32], count_comparisons: bool) -> Vec<i32> {
        let mut result = Vec::with_capacity(left.len() + right.len());
        let mut i = 0;
        let mut j = 0;

        while i < left.len() && j < right.len() {
            if count_comparisons {
                self.increment_comparisons();
            }

            if left[i] <= right[j] {
                result.push(left[i]);
                i += 1;
            } else {
                result.push(right[j]);
                j += 1;
            }
        }

        result.extend_from_slice(&left[i..]);
        result.extend_from_slice(&right[j..]);

        result
    }

    /// Sequential merge sort for base cases
    pub fn sequential_merge_sort(&self, arr: &[i32]) -> Vec<i32> {
        if arr.len() <= 1 {
            return arr.to_vec();
        }

        let mid = arr.len() / 2;
        let left = self.sequential_merge_sort(&arr[..mid]);
        let right = self.sequential_merge_sort(&arr[mid..]);

        self.merge(&left, &right, true)
    }

    /// Parallel merge sort using rayon's join
    pub fn parallel_merge_sort_rayon(&self, arr: &[i32]) -> Vec<i32> {
        if arr.len() <= self.sequential_threshold {
            return self.sequential_merge_sort(arr);
        }

        let mid = arr.len() / 2;

        // Rayon's join runs two closures potentially in parallel
        let (left, right) = rayon::join(
            || self.parallel_merge_sort_rayon(&arr[..mid]),
            || self.parallel_merge_sort_rayon(&arr[mid..]),
        );

        self.merge(&left, &right, true)
    }

    /// Parallel merge sort using rayon's parallel iterators
    pub fn parallel_merge_sort_iter(&self, arr: &[i32]) -> Vec<i32> {
        if arr.len() <= self.sequential_threshold {
            return self.sequential_merge_sort(arr);
        }

        // Split into chunks based on thread count
        let num_threads = rayon::current_num_threads();
        let chunk_size = (arr.len() + num_threads - 1) / num_threads;
        let chunk_size = chunk_size.max(self.sequential_threshold);

        // Sort chunks in parallel
        let mut sorted_chunks: Vec<Vec<i32>> = arr
            .par_chunks(chunk_size)
            .map(|chunk| self.sequential_merge_sort(chunk))
            .collect();

        // Merge sorted chunks
        while sorted_chunks.len() > 1 {
            sorted_chunks = sorted_chunks
                .par_chunks(2)
                .map(|pair| {
                    if pair.len() == 2 {
                        self.merge(&pair[0], &pair[1], false)
                    } else {
                        pair[0].clone()
                    }
                })
                .collect();
        }

        sorted_chunks.into_iter().next().unwrap_or_default()
    }

    /// Adaptive parallel merge sort
    pub fn adaptive_parallel_merge_sort(&self, arr: &[i32]) -> Vec<i32> {
        let n = arr.len();

        if n <= self.sequential_threshold {
            self.sequential_merge_sort(arr)
        } else if n <= 100_000 {
            self.parallel_merge_sort_rayon(arr)
        } else {
            self.parallel_merge_sort_iter(arr)
        }
    }

    /// Sort with performance metrics
    pub fn sort_with_metrics(&self, arr: &[i32], method: &str) -> SortResult {
        self.reset_comparisons();

        let num_threads = rayon::current_num_threads();
        let start = Instant::now();

        let sorted = match method {
            "sequential" => self.sequential_merge_sort(arr),
            "rayon_join" => self.parallel_merge_sort_rayon(arr),
            "rayon_iter" => self.parallel_merge_sort_iter(arr),
            "adaptive" => self.adaptive_parallel_merge_sort(arr),
            _ => self.adaptive_parallel_merge_sort(arr),
        };

        let time_taken = start.elapsed();

        SortResult {
            sorted_array: sorted,
            time_taken,
            comparisons: self.get_comparisons(),
            method: method.to_string(),
            num_threads,
        }
    }
}

/// Benchmark all sorting methods
pub fn benchmark_all_methods(arr: &[i32]) -> Vec<(String, SortResult)> {
    let methods = vec!["sequential", "rayon_join", "rayon_iter", "adaptive"];
    let mut results = Vec::new();

    for method in methods {
        let sorter = ParallelMergeSort::new(1000);
        let result = sorter.sort_with_metrics(arr, method);
        results.push((method.to_string(), result));
    }

    results
}

/// Calculate speedup factor
pub fn calculate_speedup(seq_time: std::time::Duration, parallel_time: std::time::Duration) -> f64 {
    if parallel_time.as_secs_f64() == 0.0 {
        return 0.0;
    }
    seq_time.as_secs_f64() / parallel_time.as_secs_f64()
}

/// Calculate parallel efficiency
pub fn calculate_efficiency(speedup: f64, num_threads: usize) -> f64 {
    if num_threads == 0 {
        return 0.0;
    }
    speedup / num_threads as f64
}

/// Check if array is sorted
pub fn is_sorted(arr: &[i32]) -> bool {
    arr.windows(2).all(|w| w[0] <= w[1])
}

fn main() {
    println!("{}", "=".repeat(80));
    println!("PARALLEL MERGE SORT DEMONSTRATION (Rust + Rayon)");
    println!("{}", "=".repeat(80));
    println!("Number of threads available: {}", rayon::current_num_threads());

    // Test with different sizes
    let sizes = vec![1000, 10000, 100000];

    for &size in &sizes {
        println!("\n{}", "=".repeat(80));
        println!("Testing with {} elements", size);
        println!("{}", "=".repeat(80));

        // Generate random array
        use rand::Rng;
        let mut rng = rand::thread_rng();
        let arr: Vec<i32> = (0..size).map(|_| rng.gen_range(0..100000)).collect();

        // Benchmark all methods
        let results = benchmark_all_methods(&arr);

        // Display results
        println!("\n{:<15} {:<15} {:<20} {:<10}", "Method", "Time", "Comparisons", "Verified");
        println!("{}", "-".repeat(80));

        let seq_time = results.iter()
            .find(|(method, _)| method == "sequential")
            .map(|(_, result)| result.time_taken)
            .unwrap_or_default();

        for (method, result) in &results {
            let verified = if is_sorted(&result.sorted_array) { "✓" } else { "✗" };

            println!(
                "{:<15} {:>13.6?}  {:>18}  {:<10}",
                method, result.time_taken, result.comparisons, verified
            );

            if method != "sequential" && seq_time.as_secs_f64() > 0.0 {
                let speedup = calculate_speedup(seq_time, result.time_taken);
                let efficiency = calculate_efficiency(speedup, result.num_threads);
                println!(
                    "{:<15} Speedup: {:.2}x  Efficiency: {:.2}%",
                    "", speedup, efficiency * 100.0
                );
            }
        }
    }

    println!("\n{}", "=".repeat(80));
    println!("RAYON CONFIGURATION ANALYSIS");
    println!("{}", "=".repeat(80));

    // Test with different thread counts
    println!("\nScalability test (50,000 elements):");
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let test_arr: Vec<i32> = (0..50000).map(|_| rng.gen_range(0..100000)).collect();

    println!("\n{:<10} {:<15} {:<10}", "Threads", "Time", "Speedup");
    println!("{}", "-".repeat(40));

    let max_threads = rayon::current_num_threads();
    let mut baseline_time = std::time::Duration::default();

    for num_threads in 1..=max_threads {
        let pool = rayon::ThreadPoolBuilder::new()
            .num_threads(num_threads)
            .build()
            .unwrap();

        let result = pool.install(|| {
            let sorter = ParallelMergeSort::new(1000);
            sorter.sort_with_metrics(&test_arr, "rayon_join")
        });

        if num_threads == 1 {
            baseline_time = result.time_taken;
        }

        let speedup = calculate_speedup(baseline_time, result.time_taken);
        println!(
            "{:<10} {:>13.6?}  {:.2}x",
            num_threads, result.time_taken, speedup
        );
    }

    println!("\n{}", "=".repeat(80));
    println!("WHEN TO USE PARALLEL MERGE SORT IN RUST");
    println!("{}", "=".repeat(80));
    println!(r#"
Parallel merge sort with Rayon is beneficial when:

✓ Dataset size > 10,000 elements
✓ Multiple CPU cores available
✓ CPU-bound comparison operations
✓ Need for memory safety without GC overhead

Rust + Rayon advantages:
✓ Zero-cost abstractions
✓ Memory safety guarantees at compile time
✓ No data races by design (borrow checker)
✓ Work-stealing scheduler (same as Go's)
✓ Excellent cache locality
✓ No runtime overhead

Avoid parallel merge sort when:

✗ Small datasets (<1,000 elements)
✗ Single-core systems
✗ Simple comparisons with negligible CPU cost
✗ Already using all cores for other tasks

Strategy selection:
- Sequential: < 1,000 elements
- Rayon Join: 1,000 - 100,000 elements (best for recursive)
- Rayon Iter: > 100,000 elements (better chunking)
- Adaptive: Let the algorithm decide

Performance characteristics:
- Thread spawn: ~1-2μs overhead
- Work stealing: Highly efficient load balancing
- Near-linear speedup: 3.5-3.8x on 4 cores
- Efficiency: 85-95% for optimal workload
- Memory overhead: Minimal (stack allocation)

Best practices:
1. Use rayon::join for divide-and-conquer algorithms
2. Use par_iter for data parallelism over collections
3. Configure thread pool size with ThreadPoolBuilder
4. Profile with cargo flamegraph
5. Use --release flag for realistic benchmarks
6. Consider cache effects (prefer smaller chunks)

Rayon features used:
- rayon::join: Fork-join parallelism
- par_chunks: Parallel iterator over chunks
- ThreadPoolBuilder: Custom thread pool configuration
- current_num_threads: Query available parallelism

Memory safety:
✓ No race conditions (enforced by borrow checker)
✓ No data races (Send + Sync traits)
✓ No use-after-free
✓ No null pointer dereferences
✓ No iterator invalidation
"#);

    println!("\nDemonstration complete!");
}

// Note: To run this code, you need to add these dependencies to Cargo.toml:
// [dependencies]
// rayon = "1.7"
// rand = "0.8"
